### PR TITLE
QT6: Update to 6.2.4, add qtcharts qtmultimedia qtshadertools

### PR DESCRIPTION
**Pull request by jhoyt4**
#### Description

Update to latest QT 6.2 LTS (6.2.4)
- Add new subports for qtcharts qtmultimedia qtshadertools

[Trac 66019](https://trac.macports.org/ticket/66019)
[Trac 66020](https://trac.macports.org/ticket/66020)
[Trac 65030](https://trac.macports.org/ticket/65030)

###### Type(s)

- [ ] bugfix
- [X] enhancement
- [ ] security fix

###### Tested on

macOS 13.0 22A380 arm64
Xcode 14.1 14B47b

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [X] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by jhoyt4** [comment]
Was hoping to get qtwebengine in this PR but am having issued where it doesn't see QtQuick despite qtdeclarative building correctly and other qt submodules seeing QtQuick.

Happy to have help getting qtwebengine working.

**Comment by jhoyt4** [comment]
Also - I squashed down to my two main commits: updating to 6.2.4 and adding new subports.  I can squash further, but thought those two were separate enough events to warrant their own commit messages.

**Comment by tenzap** [comment]
Great! This would close https://trac.macports.org/ticket/65030 too.

Just wondering, why not update to qt 6.4.0?

**Comment by tenzap** [comment]
Would you also add the new components to `./_resources/port1.0/group/qt6_variables-1.0.tcl` and do the same for all other components that are missing in it (qtdeclarative, qtsvg...) ? This would fix https://trac.macports.org/ticket/66020 too.

**Comment by jhoyt4** [comment]
> Just wondering, why not update to qt 6.4.0?

6.2 is a LTS, 6.4 [deprecates in March](https://endoflife.date/qt)

**Comment by tenzap** [comment]
> 6.2 is a LTS, 6.4 [deprecates in March](https://endoflife.date/qt)

It may be LTS but IIUC only for those with commercial license https://wiki.qt.io/Qt_6.2_Release#Qt_6.2_LTS_Commercial_Releases. Right now opensource license is stuck on 6.2.4 for at least 1 year if Qt uses the same strategy as for qt 5.15

**Comment by jhoyt4** [comment]
> 6.4 https://wiki.qt.io/Qt_6.4_Release#Qt_6.4_release_plan should be supported until the release of 6.5 LTS

Which is in [March](https://wiki.qt.io/Qt_6.5_Release) and it would make sense for someone to rev this set of ports then.

**Comment by jhoyt4** [comment]
> Would you also add the new components to `./_resources/port1.0/group/qt6_variables-1.0.tcl` and do the same for all other components that are missing in it (qtdeclarative, qtsvg...) ? This would fix https://trac.macports.org/ticket/66020 too.

@jmroot - II see you're the original committer for qt6_variables.  I'm trying to update it as @tenzap requests.  How do you handle a subport like qtdeclarative which provides multiple modules?  qtdeclarative adds:

-   lib/cmake/Qt6LabsFolderListModel
-   lib/cmake/Qt6LabsQmlModels
-   lib/cmake/Qt6LabsSettings
-   lib/cmake/Qt6PacketProtocolPrivate
-   lib/cmake/Qt6Qml
-   lib/cmake/Qt6QmlCompilerPrivate
-   lib/cmake/Qt6QmlCore
-   lib/cmake/Qt6QmlDebugPrivate
-   lib/cmake/Qt6QmlDevToolsPrivate
-   lib/cmake/Qt6QmlDomPrivate
-   lib/cmake/Qt6QmlImportScanner
-   lib/cmake/Qt6QmlLocalStorage
-   lib/cmake/Qt6QmlModels
-   lib/cmake/Qt6QmlTools
-   lib/cmake/Qt6QmlWorkerScript
-   lib/cmake/Qt6QmlXmlListModel

**Comment by tenzap** [comment]
The value seems to be used here to build a path dependency.

https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/qt6-1.0.tcl#L122

So maybe one of them would be suffficient? Like 
lib/cmake/Qt6Qml

In the qt5 equivalent there is only one line.

**Comment by jhoyt4** [comment]
Thanks!  It looks like I can also add them in as a list - so will give that a try first, then fallback to a single entry if it chokes.

**Comment by jhoyt4** [comment]
nope - chokes on the list, so single entry it is.  Looking deeper into the code, it appears mostly to just point macports dependency checking to the right place.

**Comment by jhoyt4** [comment]
That last update squashed all commits and added the request to update the qt6_variables

**Comment by tenzap** [comment]
I think you should add `Closes: https://...` lines for the 3 issues it closes in the commit message. Like below (see https://trac.macports.org/wiki/CommitMessages)

```
Closes: https://trac.macports.org/ticket/65030
Closes: https://trac.macports.org/ticket/66019
Closes: https://trac.macports.org/ticket/66202
```

**Comment by tenzap** [comment]
Did the change in `qt6_variables-1.0.tcl` actually fix your problem with `qtwebengine`?

**Comment by jhoyt4** [comment]
> Did the change in `qt6_variables-1.0.tcl` actually fix your problem with `qtwebengine`?

Somewhat - the configure script now fails vs skipping over modules. This same issue occurs with a lof of the other submodules not posted in the currrent PR and is related to qtdeclarative.

```
CMake Error at src/import/CMakeLists.txt:5 (qt_internal_add_qml_module):
  Unknown CMake command "qt_internal_add_qml_module".
```
qt_internal_add_qml_module is installed via qtdeclarative via QtQML - but the build process can't seem to find everything.  This is the same issue that caused me to keep qtmultimedia back in the original qt6 Portfile.

I believe getting those qt submodules working can are better for another PR as this one already does a lot.

For documentation purposes, here's an initial setup for the missing modules (minus qtdoc - that one's a beast...)

```
    qtactiveqt {
        {
        599d61a1856580e1160fce40209b32068aa5bed7 \
            486c1f2ccf0e545c68ebb5e139456f924b9e0948bdd16351701d624172e539f3
            276248
        }
        ""
        ""
        "qtbase"
        {"Qt5 ActiveQT"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }
    qtdatavis3d {
        {
            ae0d4f1cf69f4febfb7ab6e1f11228cab7e4af94 \
            227e170c5f0f8aa2a18cf64185fc20a730dbfe6b8a1bc23dd8407d0f8071daa0
            5109964
        }
        ""
        ""
        "qtbase qtdeclarative qtmultimedia"
        {"Pprovides multiple graph types to visualize data in 3D space"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }
    qtlottie {
        {
            e621e9d10df34f3b5055dd0ac50dbbe3d353eff9 \
            299b8070ed931060ef8ff72dbb453666d12654ddb065f8f3a6f21a1c03a4f3f2
            92220
        }
        ""
        ""
        "qtbase qtdeclarative"
        {"Lottie is a family of player software for a certain json-based file format for describing 2d vector graphics animations"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }
    qtquicktimeline {
        {
            d6ce7a9339efbf91c1dcf77c6b0fdb2171f4b041 \
            34f0622a3f7725b287d9f4cb0862ab1583de1b7046f405fb519d4d066171d9c9
            113668
        }
        ""
        ""
        "qtbase qtdeclarative"
        {"LQT Module for keyframe-based timeline construction"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }
    qtscxml {
        {
            5c43940d300c887524b3776ebe47560cfb6c6f45 \
            9b98f43eaf72c282f894564e84e1d93095e654350cb83d85f7d018b7e9efb876
            1786816
        }
        ""
        ""
        "qtbase qtdeclarative"
        {"SCXML (state machine notation) compiler and related tools"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }
    qtvirtualkeyboard {
        {
            b5c91e9638e4d705590f560f87afb65c4daa5855 \
            518a3ef2f83bac91d5997b2dd4f79fcbe436497383672681fdcc32653bb9d59b
            3744564
        }
        ""
        ""
        "qtbase qtdeclarative qtmultimedia qtsvg"
        {"QtQuick virtual keyboard"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }

    qtwebengine {
        {
            3cd272f8be2230f8d480c9b0e4642b5de1200c64 \
            2e3e96b7a0335b1f683fecf8863ff2930b1bed4df00ce2ff064fa26d0a49e2b1
            369689936
        }
        "port:python310 port:py310-ply path:bin/ninja:ninja port:gperf port:bison port:flex"
        "port:fontconfig port:dbus path:lib/pkgconfig/harfbuzz.pc:harfbuzz path:lib/pkgconfig/glib-2.0.pc:glib2 port:zlib port:minizip port:libevent port:libxml2 port:jsoncpp port:protobuf3-cpp path:lib/pkgconfig/poppler.pc:poppler port:pulseaudio path:lib/pkgconfig/icu-uc.pc:icu path:lib/libavcodec.dylib:ffmpeg port:libopus port:webp port:libpng port:lcms2 port:freetype port:re2 port:snappy"
        "qtdeclarative qtpositioning qttools qtwebchannel"
        {"Qt WebEngine"}
        "very large and relatively new"
        "variant overrides: "
        "revision 0"
        "License: "
    }
    qtwebview {
        {
            3c1490d7d3b6f53c578e9b34696a4496cbb67c66 \
            407de86a9b6ae95621f2e976895b62809b6258f9f528179798d232d62b74b34f
            136656
        }
        ""
        ""
        "qtdeclarative qtwebengine"
        {"Qt Web View"}
        ""
        "variant overrides: "
        "revision 0"
        "License: "
    }
```

**Comment by jhoyt4** [comment]
I believe a part of the issue is how far along the submodules are in converting from qmake to "qt-configure-module".  I'm hoping when qt6.5 releases next year, more submodules will be fully updated.

**Comment by tenzap** [comment]
Debian is able to build it since 6.2.4 but they configure cmake instead of "qt-configure-module".

For example

`cd obj-x86_64-linux-gnu && cmake -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON -DFETCHCONTENT_FULLY_DISCONNECTED=ON -DCMAKE_INSTALL_RUNSTATEDIR=/run -DCMAKE_SKIP_INSTALL_ALL_DEPENDENCY=ON -GNinja -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_INSTALL_LIBDIR=lib/x86_64-linux-gnu -DCMAKE_LIBRARY_PATH=x86_64-linux-gnu -DFEATURE_webengine_system_icu=ON -DFEATURE_webengine_system_libevent=ON -DFEATURE_qtpdf_build=ON -DFEATURE_qtpdf_widgets_build=ON -DFEATURE_qtpdf_quick_build=ON -DFEATURE_webengine_proprietary_codecs=ON -DFEATURE_webengine_kerberos=ON ..`

**Comment by jhoyt4** [comment]
There were a ton of issue trying to build this a year ago with cmake.  I'm sure these select submodules could be built that way.  Happy to have someone else give it a go in another PR :)

**Comment by tenzap** [comment]
Thanks for your work. Hopefully this will fix my issues with welle.io. Eager to see the result :)

**Comment by tenzap** [comment]
BTW, for the close I meant "commit message", not summary of the PR.

**Comment by jhoyt4** [comment]
Thanks - need another cup of coffee this AM.  Should now be fixed in that last force push.

**Comment by reneeotten** [comment]
I have tried this before and it will need a lot of work to get this all done correctly.... The main issue is as @jhoyt4 says with `qt_internal_add_qml_module` - see also [here](https://forum.qt.io/topic/136134/building-of-certain-modules-fails-with-unknown-cmake-command-qt_internal_add_qml_module/3). 

I haven't been able to resolve this and left my work-in-progress branch [here](https://github.com/reneeotten/macports-ports/commits/qt6). I did take a quick look yesterday night but with the separation of the original qt6 PG into two parts this needs additional work to be able to rebase against current `master`. I don't think I'll have a chance to do anything on this anytime soon, but perhaps there is something useful in there for someone else.

**Comment by jhoyt4** [comment]
@reneeotten - thanks for the pointers and sanity check!

I believe any further updates should be in a different PR as this one achieves the goal of updating the ports to the latest LTS and even went into OT to add some new ports.

**Comment by jhoyt4** [comment]
@reneeotten - I have a merged version of your wip branch and this one on my local machine and am compiling it currently.

I noticed that you added yourself in your repo as the maintainer.  Did you want me to make that update here as well?

**Comment by reneeotten** [comment]
> @reneeotten - I have a merged version of your wip branch and this one on my local machine and am compiling it currently.
> 
> 
> 
> I noticed that you added yourself in your repo as the maintainer.  Did you want me to make that update here as well?

this should retain my separate commits and not be lumped into one big change. Also, last time I checked it actually doesn't build all the modules so this cannot be merged until that is fixed and thoroughly tested.

I don't need to be the maintainer or can co-maintain with you if you want, but regardless this should be fixed completely before merging. Main point is properly building of qtdeclarative so that other ports can find stuff and don't fail anymore.

**Comment by jhoyt4** [comment]
> this should retain my separate commits and not be lumped into one big change. Also, last time I checked it actually doesn't build all the modules so this cannot be merged until that is fixed and thoroughly tested.

ok - I'll have to roll back that last squashed commit to break them out again.

I left the non-functional modules out until a fix can be made.
 
> I don't need to be the maintainer or can co-maintain with you if you want, but regardless this should be fixed completely before merging. Main point is properly building of qtdeclarative so that other ports can find stuff and don't fail anymore.

I'll leave the maintainer as nomaintainer.

Understood - I still believe getting something out that now includes qt 6.2.4 and the three new ports should happen.

...time to break up that last commit.

**Comment by jhoyt4** [comment]
That last force push included only my changes for 6.2.4, the new ports, and at I added to the variables

**Comment by jhoyt4** [comment]
@reneeotten - I'm going to leave your changes for another PR.  Once the CI's finish, this should be ready to merge.

**Comment by jhoyt4** [comment]
darn it - forgot to remove the SDK 12 now that its no longer needed...  Hopefully this last one gets it going

File: _resources/port1.0/group/qt6_variables-1.0.tcl [hunk 1]
```diff
@@ -118,38 +118,140 @@ namespace eval qt6pg {
     #
     ###############################################################################
     array set qt6_component_lib {
+        qt3d {
```
Comment (reneeotten) [inline _resources/port1.0/group/qt6_variables-1.0.tcl:121]: do all the Qt modules that you added here build? If not, they probably shouldn't be added now as they will show up as subports for the qt6 port - correct?

Comment (jhoyt4) [reply _resources/port1.0/group/qt6_variables-1.0.tcl:121]: Yes they do.  But I forgot to remove qtlocation from the portfile, so will do that shortly.

Comment (jhoyt4) [reply _resources/port1.0/group/qt6_variables-1.0.tcl:121]: That last force push removed qtlocation.

Comment (reneeotten) [reply _resources/port1.0/group/qt6_variables-1.0.tcl:121]: okay, sorry I might not have looked carefully enough... I just upgrade to Ventura so am building a lot of stuff before I can get to this PR to do some local testing.... Once I've done that I'll either merge this PR or get back here with comments. Thanks @jhoyt4 for working on this!

**Comment by jhoyt4** [comment]
darn it - it was working before I rolled in all the other stuff.

**Comment by jhoyt4** [comment]
it helps to include qtmultimedia since its working...

**Comment by jhoyt4** [comment]
ok - I just rebuilt everything on my machine (Venture / M1 mini) and it worked.  Hopefully nothing new turns up in the CI process.

**Comment by jhoyt4** [comment]
That last push was just to rebase to the latest master since qt6 had an update yesterday.

**Comment by jmroot** [comment]
> @jmroot - II see you're the original committer for qt6_variables.

I just broke the qt6 portgroup up into multiple portgroups, because it was relying on undefined behaviour of multiple inclusion of the same portgroup, which would have made it break under MacPorts 2.8. I'm not at all familiar with the packaging of Qt 6.

**Comment by jhoyt4** [comment]
That last push was another rebase.  This PR is ready for review and commit.

**Comment by reneeotten** [comment]
thanks @jhoyt4! I just got around testing this PR and it all builds for me as well, so let's merge this and we can fine-tune some of the things I had sitting in a local branch at some later point.

**Comment by jhoyt4** [comment]
@reneeotten sounds good.  It may make sense to try again in a few months when the next LTS drops (6.5).  Looking at the submodules, it looks like they are slowly migrating to the QT6 way of things - possibly some of the "qt_internal_add_qml_module" issues will get resolved upstream.

**Comment by tenzap** [comment]
Good but still not usable by qml protects. 
See #16831

**Pull request merged**